File: selection/selection_parallel_new.jl
using Base.Threads
using Polyester

function _selection_ranges(lo::Int, hi::Int, parts::Int)
    hi < lo && return UnitRange{Int}[]
    total = hi - lo + 1
    actual = min(max(parts, 1), total)
    ranges = Vector{UnitRange{Int}}(undef, actual)
    base, extra = divrem(total, actual)
    start = lo

    for idx in 1:actual
        len = base + (idx <= extra ? 1 : 0)
        stop = start + len - 1
        ranges[idx] = start:stop
        start = stop + 1
    end

    return ranges
end

function parallel_selection_sort!(arr::Vector)
    n = length(arr)
    n < 2 && return arr

    for i in 1:(n - 1)
        ranges = _selection_ranges(i, n, Threads.nthreads())
        local_min = fill(i, length(ranges))

        @batch for ridx in eachindex(ranges)
            min_idx = first(ranges[ridx])
            for j in ranges[ridx]
                if arr[j] < arr[min_idx]
                    min_idx = j
                end
            end
            local_min[ridx] = min_idx
        end

        min_idx = local_min[1]
        for idx in local_min
            if arr[idx] < arr[min_idx]
                min_idx = idx
            end
        end

        if min_idx != i
            arr[i], arr[min_idx] = arr[min_idx], arr[i]
        end
    end

    return arr
end

File: selection/selection_parallel_new.py
from parallel_utils import parallel_ranges
from concurrent.futures import ThreadPoolExecutor


def parallel_selection_sort(arr, left=0, right=None):
    if right is None:
        right = len(arr) - 1
    if left >= right:
        return arr

    for i in range(left, right):
        ranges = parallel_ranges(right - i + 1)

        def local_min(index_range):
            min_idx = i
            for offset in index_range:
                j = i + offset
                if arr[j] < arr[min_idx]:
                    min_idx = j
            return min_idx

        if len(ranges) == 1:
            candidates = [local_min(ranges[0])]
        else:
            with ThreadPoolExecutor(max_workers=len(ranges)) as executor:
                candidates = list(executor.map(local_min, ranges))

        min_idx = min(candidates, key=lambda idx: (arr[idx], idx))
        if min_idx != i:
            arr[i], arr[min_idx] = arr[min_idx], arr[i]

    return arr
